lib/PuppeteerSharp.Tests/Frame/WaitForSelectorTests.cs

```diff
@@ -0,0 +1,216 @@
+            var frame = Page.MainFrame;
+            var added = false;
+            var watchdog = frame.WaitForSelectorAsync("div").ContinueWith(_ => added = true);
+            // run nop function..
+            await frame.EvaluateExpressionAsync("42");
+            // .. to be sure that waitForSelector promise is not resolved yet.
+            Assert.False(added);
+
+            await frame.EvaluateFunctionAsync(AddElement, "br");
+            Assert.False(added);
+
+            await frame.EvaluateFunctionAsync(AddElement, "div");
+            await watchdog;
+            Assert.True(added);
+        }
+
+        [Fact]
+        public async Task ShouldWorkWhenNodeIsAddedThroughInnerHTML()
+        {
+            await Page.GoToAsync(TestConstants.EmptyPage);
+            var watchdog = Page.WaitForSelectorAsync("h3 div");
+            await Page.EvaluateFunctionAsync(AddElement, "span");
+            await Page.EvaluateExpressionAsync("document.querySelector('span').innerHTML = '<h3><div></div></h3>'");
+            await watchdog;
+        }
+
+        [Fact]
+        public async Task PageWaitForSelectorAsyncIsShortcutForMainFrame()
+        {
+            await Page.GoToAsync(TestConstants.EmptyPage);
+            await FrameUtils.AttachFrame(Page, "frame1", TestConstants.EmptyPage);
+            var otherFrame = Page.Frames.ElementAt(1);
+            var added = false;
+            Page.WaitForSelectorAsync("div").ContinueWith(_ => added = true);
+            await otherFrame.EvaluateFunctionAsync(AddElement, "div");
+            Assert.False(added);
+
+            await Page.EvaluateFunctionAsync(AddElement, "div");
+            Assert.True(added);
+        }
+
+        [Fact]
+        public async Task ShouldRunInSpecifiedFrame()
+        {
+            await FrameUtils.AttachFrame(Page, "frame1", TestConstants.EmptyPage);
+            await FrameUtils.AttachFrame(Page, "frame2", TestConstants.EmptyPage);
+            var frame1 = Page.Frames.ElementAt(1);
+            var frame2 = Page.Frames.ElementAt(2);
+            var added = false;
+            frame2.WaitForSelectorAsync("div").ContinueWith(_ => added = true);
+            Assert.False(added);
+
+            await frame1.EvaluateFunctionAsync(AddElement, "div");
+            Assert.False(added);
+
+            await frame2.EvaluateFunctionAsync(AddElement, "div");
+            Assert.True(added);
+        }
+
+        [Fact]
+        public async Task ShouldThrowIfEvaluationFailed()
+        {
+            await Page.EvaluateOnNewDocumentAsync(@"function() {
+                document.querySelector = null;
+            }");
+            await Page.GoToAsync(TestConstants.EmptyPage);
+            var exception = await Assert.ThrowsAnyAsync<PuppeteerException>(() => Page.WaitForSelectorAsync("*"));
+            Assert.Contains("document.querySelector is not a function", exception.Message);
+        }
+
+        [Fact(Skip = "FrameUtils.DetachFrame hangs :(")]
```

Comment:
Do you need a hand here?

Reply:
Not sure yet. I didn't give a lot of time. I just figured out what the problem